### DISCUSSION TITLE
GLSurfaceView onTouchEvent not working at all.

### DISCUSSION THREAD

**Discussion by jupiterspot** [community/community#21383] [Other Feature Feedback, Questions, & Ideas]
<p>I have a simple Android application with OpenGL (GLES Version 2) that draws a simple colored square and works without a problem so far. I have the GlSurfaceView set as my own custom View and the Renderer where the square is drawn.</p>
<p>I am confused with the onTouchEvent. I want to change the square’s color to a random one every time I touch the screen. The process of setting up random values within a range for the square’s color array is easy to do, but getting the touch listener to work appears not.</p>
<p>So far, I’ve been only getting results online with outdated GLES versions or the solutions I’m looking for, but they don’t seem to work for me at all. First thing I tried to do is overriding the onTouchEvent in my GLSurfaceView class where my custom function in the Renderer gets called to change the square’s color when there’s a touch event, in this case when I simply tap on my screen. Calling the <span class="mention">@Override</span> onTouchEvent in GLSurfaceView didn’t do anything. Then I tried setting the touch listener for the GLSurfaceView view in the Main Activity. Still nothing. It seems to be only not working for me for some reason.</p>
<p>Here’s what I did. (Currently you’re seeing setting the listener for the GLSurfaceView view in the MainActivity):</p>
<p><strong>activity_main.xml</strong></p>
<p></p><pre>&lt;FrameLayout xmlns:android=“<a href="http://schemas.android.com/apk/res/android" class="inline-onebox-loading" rel="nofollow noopener">http://schemas.android.com/apk/res/android</a>”<br>
xmlns:app=“<a href="http://schemas.android.com/apk/res-auto" class="inline-onebox-loading" rel="nofollow noopener">http://schemas.android.com/apk/res-auto</a>”<br>
xmlns:tools=“<a href="http://schemas.android.com/tools" class="inline-onebox-loading" rel="nofollow noopener">http://schemas.android.com/tools</a>”<br>
android:layout_width=“match_parent”<br>
android:layout_height=“match_parent”<br>
tools:context=".MainActivity"&gt;<p></p>
<p><strong> &lt;com.example.opengltest.OpenGLView<br>
android:id="@+id/openGLView"<br>
android:layout_width=“match_parent”<br>
android:layout_height=“match_parent”/&gt;</strong></p>
<p>&lt;/FrameLayout&gt;</p></pre><p></p>
<p><strong>MainActivity.java</strong></p>
<p></p><pre>public class MainActivity extends AppCompatActivity {<p></p>
<pre><code>private OpenGLView openGLView;

@Override
protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    openGLView = (OpenGLView) findViewById(R.id.openGLView);
    openGLView.setOnTouchListener(new View.OnTouchListener() {

        @Override
        public boolean onTouch(View v, MotionEvent e)
        {
            OpenGLRenderer mRenderer = new OpenGLRenderer();

            if(e.getAction() == MotionEvent.ACTION_DOWN)
            {
                mRenderer.changeColor();
            }

            return true;
        }
    });
}

@Override
protected void onResume() {
    super.onResume();
    openGLView.onResume();
}

@Override
protected void onPause() {
    super.onPause();
    openGLView.onPause();
}
</code></pre>
<p>}</p></pre><p></p>
<p><strong>OpenGLView.java</strong> (This is my GLSurfaceView class)</p>
<p></p><pre>public class OpenGLView extends GLSurfaceView<br>
{<br>
public OpenGLView(Context context) {<br>
super(context);<br>
init();<br>
}<p></p>
<pre><code>public OpenGLView(Context context, AttributeSet attrs) {
    super(context, attrs);
    init();
}

private void init()
{
    setEGLContextClientVersion(2);
    setPreserveEGLContextOnPause(true);
    setRenderer(new OpenGLRenderer());
}
</code></pre>
<p>}</p></pre><p></p>
<p><strong>OpenGLRenderer.java</strong> (My renderer class)</p>
<p></p><pre>public class OpenGLRenderer implements GLSurfaceView.Renderer<br>
{<br>
//… onSurfaceChanged, onSurfaceCreated, onDrawFrame, …<p></p>
<pre><code>// This is the function called to change the square's color. I tried putting it inside onDrawFrame to change the color on every new frame and it works without a problem.
public void changeColor()
{
    mySquareColorData[0] = (float) ThreadLocalRandom.current().nextDouble(0.0, 1.0);
    mySquareColorData[1] = (float) ThreadLocalRandom.current().nextDouble(0.0, 1.0);
    mySquareColorData[2] = (float) ThreadLocalRandom.current().nextDouble(0.0, 1.0);
}
</code></pre>
<p>}</p></pre><p></p>
<p>I sure hope I haven’t missed something. I only gave you the necessary information. Everything GLES related is working except the onTouchEvent.</p>
<p>Thanks in advance!</p>

**Comment by fire-eggs** [comment]
<p>You might check to see if there is a problem with touch mode. Try calling </p>
<pre><code>isInTouchMode()
</code></pre>
<p>in the view to confirm that touch is enabled.</p>
<p>It’s possible (see <a href="https://stackoverflow.com/questions/5024892/glsurfaceview-not-getting-ontouch-events" rel="nofollow noopener">this</a>) your onTouch override in the activity needs to read as so:</p>
<pre><code>@Override
    protected boolean onTouchEvent(MotionEvent event) {
        Log.i("motion", event);
        return true;
    }
</code></pre>
<p>Would a click listener do the trick?</p>
<p>Please follow-up to let us know how you made out. For good karma, mark a reply as the answer if it helped!</p>

**Comment by jupiterspot** [comment, marked as answer]
<p>Hello, sorry for the late reply. I have solved my problem.</p>
<p>The renderer class should implement the GLSurfaceView class that I have made instead.</p>
<pre><code>public class OpenGLRenderer implements&lt;font color="#800000"&gt;&lt;strike&gt;GLSurfaceView&lt;/strike&gt;&lt;/font&gt;&lt;font color="#339966"&gt;OpenGLView&lt;/font&gt;.Renderer
</code></pre>
